pkg/oauth/tokenrequest/basicauth.go

```diff
@@ -0,0 +1,160 @@
+	username, password string,
+) *BasicChallengeHandler {
+	return &BasicChallengeHandler{
+		Host:             host,
+		Reader:           reader,
+		Writer:           writer,
+		passwordPrompter: passwordPrompter,
+		Username:         username,
+		Password:         password,
+	}
+}
+
+type BasicChallengeHandler struct {
+	// Host is the server being authenticated to. Used only for displaying messages when prompting for username/password
+	Host string
+
+	// Reader is used to prompt for username/password. If nil, no prompting is done
+	Reader io.Reader
+	// Writer is used to output prompts. If nil, stdout is used
+	Writer io.Writer
+
+	// passwordPrompter is used to retrieve the password as a string
+	passwordPrompter PasswordPrompter
+
+	// Username is the username to use when challenged. If empty, a prompt is issued to a non-nil Reader
+	Username string
+	// Password is the password to use when challenged. If empty, a prompt is issued to a non-nil Reader
+	Password string
+
+	// handled tracks whether this handler has already handled a challenge.
+	handled bool
+	// prompted tracks whether this handler has already prompted for a username and/or password.
+	prompted bool
+}
+
+func (c *BasicChallengeHandler) CanHandle(headers http.Header) bool {
+	isBasic, _ := basicRealm(headers)
+	return isBasic
+}
+func (c *BasicChallengeHandler) HandleChallenge(requestURL string, headers http.Header) (http.Header, bool, error) {
+	if c.prompted {
+		klog.V(2).Info("already prompted for challenge, won't prompt again")
+		return nil, false, nil
+	}
+	if c.handled {
+		klog.V(2).Info("already handled basic challenge")
+		return nil, false, nil
+	}
+
+	username := c.Username
+	password := c.Password
+
+	missingUsername := len(username) == 0
+	missingPassword := len(password) == 0
+
+	if missingUsername {
+		return nil, false, NewBasicAuthNoUsernameError()
+	}
+	// Basic auth does not support usernames containing colons
+	// http://tools.ietf.org/html/rfc2617#section-2
+	if strings.Contains(username, ":") {
+		return nil, false, fmt.Errorf("username %s is invalid for basic auth", username)
+	}
+	if missingPassword && c.Reader != nil {
+		w := c.Writer
+		if w == nil {
+			w = os.Stdout
+		}
+
+		if _, realm := basicRealm(headers); len(realm) > 0 {
+			fmt.Fprintf(w, "Authentication required for %s (%s)\n", c.Host, realm)
+		} else {
+			fmt.Fprintf(w, "Authentication required for %s\n", c.Host)
+		}
+		fmt.Fprintf(w, "Username: %s\n", username)
+		if missingPassword {
+			password = c.passwordPrompter.PromptForPassword(c.Reader, w, "Password: ")
+		}
+		// remember so we don't re-prompt
+		c.prompted = true
+	}
+
+	responseHeaders := http.Header{}
+	responseHeaders.Set("Authorization", getBasicHeader(username, password))
+	// remember so we don't re-handle non-interactively
+	c.handled = true
+	return responseHeaders, true, nil
+}
+func (c *BasicChallengeHandler) CompleteChallenge(requestURL string, headers http.Header) error {
+	return nil
```

Comment:
is this implementing some interface? If so, which one?

Reply:
`ChallengeHandler`, added that to the struct's godoc